sycl/include/sycl/ext/intel/experimental/cache_control_properties.hpp

```diff
@@ -0,0 +1,249 @@
+}
+
+} // namespace detail
+
+template <cache_mode M, cache_level... Ls> struct cache_control {
+  static constexpr const auto mode = M;
+  static constexpr const int countL1 = ((Ls == cache_level::L1 ? 1 : 0) + ...);
+  static constexpr const int countL2 = ((Ls == cache_level::L2 ? 1 : 0) + ...);
+  static constexpr const int countL3 = ((Ls == cache_level::L3 ? 1 : 0) + ...);
+  static constexpr const int countL4 = ((Ls == cache_level::L4 ? 1 : 0) + ...);
+  static constexpr const uint32_t levels = ((1 << static_cast<int>(Ls)) | ...);
+  static constexpr const uint32_t encoding =
+      (countL1, countL2, countL3, countL4, detail::checkLevel1<countL1>(),
+       detail::checkLevel2<countL2>(), detail::checkLevel3<countL3>(),
+       detail::checkLevel4<countL4>(), levels << static_cast<int>(M) * 4);
+};
+
+template <typename PropertyT, typename... Ts>
+using property_value =
+    sycl::ext::oneapi::experimental::property_value<PropertyT, Ts...>;
+
+struct read_hint_key {
+  template <typename... Cs>
+  using value_t = property_value<read_hint_key, Cs...>;
+};
+
+struct read_assertion_key {
+  template <typename... Cs>
+  using value_t = property_value<read_assertion_key, Cs...>;
+};
+
+struct write_hint_key {
+  template <typename... Cs>
+  using value_t = property_value<write_hint_key, Cs...>;
+};
+
+template <typename... Cs>
+inline constexpr read_hint_key::value_t<Cs...> read_hint;
+
+template <typename... Cs>
+inline constexpr read_assertion_key::value_t<Cs...> read_assertion;
+
+template <typename... Cs>
+inline constexpr write_hint_key::value_t<Cs...> write_hint;
+
+} // namespace experimental
+} // namespace intel
+
+namespace oneapi {
+namespace experimental {
+
+template <typename T, typename PropertyListT> class annotated_ptr;
+
+template <>
+struct is_property_key<intel::experimental::read_hint_key> : std::true_type {};
+template <typename T, typename PropertyListT>
+struct is_property_key_of<intel::experimental::read_hint_key,
+                          annotated_ptr<T, PropertyListT>> : std::true_type {};
+
+template <>
+struct is_property_key<intel::experimental::read_assertion_key>
+    : std::true_type {};
+template <typename T, typename PropertyListT>
+struct is_property_key_of<intel::experimental::read_assertion_key,
+                          annotated_ptr<T, PropertyListT>> : std::true_type {};
+
+template <>
+struct is_property_key<intel::experimental::write_hint_key> : std::true_type {};
+template <typename T, typename PropertyListT>
+struct is_property_key_of<intel::experimental::write_hint_key,
+                          annotated_ptr<T, PropertyListT>> : std::true_type {};
+
+namespace detail {
+
+// Values assigned to cache levels in a nibble.
+static constexpr int L1BIT = 1;
+static constexpr int L2BIT = 2;
+static constexpr int L3BIT = 4;
+static constexpr int L4BIT = 8;
+
+// Starting bit position for cache levels of a cache mode.
+static constexpr int UNCACHED = 0;
```

Comment:
I do not see its being used somewhere. Could you please point me to where it is used/going to be used?

Reply:
"UNCACHED" is not used. I could use it and the compiler will optimize the use away. Or I could remove it. I kept it in there more as documentation.

Reply:
Ok

Reply:
I have removed the unused declarations and added a comment instead.